base/modulus_remainder.cc

```diff
@@ -0,0 +1,232 @@
+#include "base/modulus_remainder.h"
+
+// Based on https://github.com/halide/Halide/blob/main/src/ModulusRemainder.cpp.
+namespace slinky {
+
+namespace {
+
+int64_t div_imp(int64_t a, int64_t b) {
+  int64_t ia = a;
+  int64_t ib = b;
+  int64_t a_neg = ia >> 63;
+  int64_t b_neg = ib >> 63;
+  int64_t b_zero = (ib == 0) ? -1 : 0;
+  ib -= b_zero;
+  ia -= a_neg;
+  int64_t q = ia / ib;
+  q += a_neg & (~b_neg - b_neg);
+  q &= ~b_zero;
+  return q;
+}
+
+// A version of mod where a % 0 == a
+int64_t mod_imp(int64_t a, int64_t b) {
+    if (b == 0) {
+        return a;
+    } else {
+      int64_t ia = a;
+      int64_t ib = b;
+      int64_t a_neg = ia >> 63;
+      int64_t b_neg = ib >> 63;
+      int64_t b_zero = (ib == 0) ? -1 : 0;
+      ia -= a_neg;
+      int64_t r = ia % (ib | b_zero);
+      r += (a_neg & ((ib ^ b_neg) + ~b_neg));
+      r &= ~b_zero;
+
+      return r;
+    }
+}
+
+int64_t gcd(int64_t a, int64_t b) {
+    if (a < b) {
```

Comment:
We have `gcd` too, but it looks like maybe this implementation is better?

Reply:
Yes, I think this one is better, I replace the one in base/.